log4j-layout-template-json/src/main/java/org/apache/logging/log4j/layout/template/json/resolver/ReadOnlyStringMapResolver.java

```diff
@@ -0,0 +1,386 @@
+ *   "flatten": {
+ *     "prefix": "_"
+ *   }
+ * }
+ * </pre>
+ *
+ * @see MapResolver
+ * @see ThreadContextDataResolver
+ */
+class ReadOnlyStringMapResolver implements EventResolver {
+
+    private final EventResolver internalResolver;
+
+    ReadOnlyStringMapResolver(
+            final EventResolverContext context,
+            final TemplateResolverConfig config,
+            final Function<LogEvent, ReadOnlyStringMap> mapAccessor) {
+        this.internalResolver = createResolver(context, config, mapAccessor);
+    }
+
+    private static EventResolver createResolver(
+            final EventResolverContext context,
+            final TemplateResolverConfig config,
+            final Function<LogEvent, ReadOnlyStringMap> mapAccessor) {
+        final Object flattenObject = config.getObject("flatten");
+        final boolean flatten;
+        if (flattenObject == null) {
+            flatten = false;
+        } else if (flattenObject instanceof Boolean) {
+            flatten = (boolean) flattenObject;
+        } else if (flattenObject instanceof Map) {
+            flatten = true;
+        } else {
+            throw new IllegalArgumentException("invalid flatten option: " + config);
+        }
+        final String key = config.getString("key");
+        final String prefix = config.getString(new String[] {"flatten", "prefix"});
+        final String pattern = config.getString("pattern");
+        final boolean stringified = config.getBoolean("stringified", false);
+        if (key != null) {
+            if (flatten) {
+                throw new IllegalArgumentException(
+                        "both key and flatten options cannot be supplied: " + config);
+            }
+            return createKeyResolver(key, stringified, mapAccessor);
+        } else {
+            final RecyclerFactory recyclerFactory = context.getRecyclerFactory();
+            return createResolver(
+                    recyclerFactory,
+                    flatten,
+                    prefix,
+                    pattern,
+                    stringified,
+                    mapAccessor);
+        }
+    }
+
+    private static EventResolver createKeyResolver(
+            final String key,
+            final boolean stringified,
+            final Function<LogEvent, ReadOnlyStringMap> mapAccessor) {
+        return new EventResolver() {
+
+            @Override
+            public boolean isResolvable(final LogEvent logEvent) {
+                final ReadOnlyStringMap map = mapAccessor.apply(logEvent);
+                return map != null && map.containsKey(key);
+            }
+
+            @Override
+            public void resolve(final LogEvent logEvent, final JsonWriter jsonWriter) {
+                final ReadOnlyStringMap map = mapAccessor.apply(logEvent);
+                final Object value = map == null ? null : map.getValue(key);
+                if (stringified) {
+                    final String valueString = String.valueOf(value);
```

Comment:
I realize this code has just moved and isn't new in this commit, but we might want to use something like `ParameterizedMessage.deepToString(value)` for better handling of standard types, and exception handling in the case toString throws.

Reply:
Good catch @carterkozak! I did not know about `ParameterizedMessage.deepToString(value)`. I will resolve this conversation to get this PR merged and address your remark in a separate ticket, [LOG4J2-2966](https://issues.apache.org/jira/browse/LOG4J2-2966).